blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,734 @@
+Using the [MethodAnalysis] package (which is at a very early stage of development
+at the time of this writing), you can find out that a fresh Julia session has almost 50,000 `MethodInstance`s tucked away in its cache.
+These are mostly for `Base` and the standard libraries.
+(There are some additional `MethodInstance`s that get created to load the MethodAnalysis package and do this analysis, but these are surely a very small fraction of the total.)
+
+Using some not-yet merged work in both Julia itself and [SnoopCompile], we can count the number of invalidations when we load various packages into a fresh Julia session:
+
+| Package | Version | # of unique invalidations |
+|:------- | -------:| ------------------:|
+| Example | 0.5.3 | 0 |
+| Revise | 2.6.6 | 27 |
+| FixedPointNumbers | 0.8.0 | 429 |
+| SIMD | 2.8.0 | 2799 |
+| StaticArrays | 0.12.3 | 2852 |
+| Optim | 0.21.0 | 3171 |
+| Images | 0.22.2 | 3638 |
+| Flux | 0.10.4 | 3697 |
+| Plots | 1.2.3 | 4002 |
+| DataFrames | 0.21.0 | 4048 |
+| JuMP | 0.21.2 | 4666 |
+| Makie | 0.10.0 | 6118 |
+| DifferentialEquations | 6.13.0 | 6777 |
+
+You can see that key packages used by large portions of the Julia ecosystem invalidate
+hundreds or thousands of `MethodInstance`s, sometimes more than 10% of the total
+number of `MethodInstance`s present before loading the package.
+
+## How serious is method invalidation?
+
+The next time you want to call functionality that gets invalidated,
+you have to wait for recompilation.
+We can illustrate this using everyone's favorite example, plotting:
+
+```julia-repl
+julia> using Plots
+
+julia> @time display(plot(rand(5)))
+  7.717729 seconds (15.27 M allocations: 797.207 MiB, 3.59% gc time)
+```
+
+As is well known, it's much faster the second time, because it's already compiled:
+
+```julia-repl
+julia> @time display(plot(rand(5)))
+  0.311226 seconds (19.93 k allocations: 775.055 KiB)
+```
+
+Moreover, if you decide you want some additional functionality and decide to load a new package, sometimes it's essentially as fast again:
+
+```julia-repl
+julia> using StaticArrays
+
+julia> @time display(plot(rand(5)))
+  0.305394 seconds (19.96 k allocations: 781.836 KiB)
+```
+
+But if you load a package that does a lot of invalidation:
+
+```julia-repl
+julia> using SIMD
+
+julia> @time display(plot(rand(5)))
+  7.238336 seconds (26.50 M allocations: 1.338 GiB, 7.88% gc time)
+```
+
+Because so much got invalidated by loading SIMD, Julia had to recompile many methods before it could once again produce a plot, so that in terms of time it was almost as expensive as the first usage.
+The size of the effect varies substantially depending on what task you are trying to achieve and what packages you load to do the invalidation.
+
+It's worth noting that you can escape much of this cost by loading all
+packages at the outset.
+Loading time is somewhat increased by invalidation, but the cost of
+first-time usage is typically larger. When possible, it's best to get
```

Comment:
> Loading time is somewhat increased by invalidation, but the cost of
> first-time usage is typically larger.

Can you clarify this line? Do you mean "the cost of first-time usage _after invalidation_"?

Reply:
Larger than loading. Invalidation can affect both, but changes to the first-time usage may have more effect.